ui/src/components/Note.tsx

```diff
@@ -1,22 +1,115 @@
-import React from 'react';
+import React, { useState, useRef } from 'react';
 
 import * as types from '../types';
 
 import './Note.scss'
 
 interface NoteProps {
-  note: types.Note;
-  showAuthor: boolean;
+  note?: types.Note;
   participants: Map<string, types.Participant>;
+  editable: boolean;
+  onNoteSave: (text: string, id?: number) => void;
+  tabIndex?: number
 }
 
-export default function noteComponent({note, showAuthor, participants}: NoteProps) {
-  return <div className='Note' key={note.id}>
-    <p>
-      {note.text}
-    </p>
-    {showAuthor && <em className="Note__Author">
-      {participants.get(note.authorId)?.name || "Unknown author"}
-    </em>}
+export default function Note({note, editable, participants, onNoteSave, tabIndex}: NoteProps) {
+  const textareaRef = useRef<HTMLTextAreaElement>(null);
+  const [editState, setEdit] = useState<{editText?: string, beingEdited?: boolean}>({editText: "", beingEdited: false})
+  const {editText, beingEdited} = editState
+  const isCreateEditor = !note
+
+  const textFromEdit = beingEdited || !note
+  const author = note ? (participants.get(note.authorId)?.name || "Unknown author") : null
+
+  // Callbacks
+
+  const handleEdit = () => {
+    setEdit({editText: note!.text, beingEdited: true})
+  }
+
+  const handleDelete = () => {
+    onNoteSave("", note?.id)
+    setEdit({beingEdited: false})
+  }
+
+  const handleCancelEdition = () => {
+    setEdit({beingEdited: false})
+  }
+
+  const handleSave = () => {
+    onNoteSave(editText!, note?.id)
+    if (isCreateEditor) {
+      // TODO: Find better solution.
+      // This is it createEditor
+      setEdit({editText: "", beingEdited: false})
+      textareaRef.current?.focus()
+    } else {
+      setEdit({...editState, beingEdited: false})
+    }
+  }
+
+  // Badges
+
+  const editBadge = () => <button onClick={handleEdit} className="Note__badge Note__edit">✎</button>
+  const authorBadge = () => <em className="Note__badge">{ author }</em>
+
+  const saveBadge = () => <button key="save" onClick={handleSave} className="Note__badge"> { note ? "✓" : "↵" } </button>
+  const cancelBadge = () => <button key="cancel" onClick={handleCancelEdition} className="Note__badge"> ␘ </button>
+  const deleteBadge = () => <button key="delete" onClick={handleDelete} className="Note__badge"> ␡ </button>
```

Comment:
those two icons don't display for me, neither in github nor in the UI :-/ They're maybe somehow just in the default macOS font?

![image](https://user-images.githubusercontent.com/2526296/82027056-f69c0380-9693-11ea-9333-3df278507fd0.png)
![image](https://user-images.githubusercontent.com/2526296/82027107-03205c00-9694-11ea-8c1a-94cc14b3ddfd.png)

Reply:
(the ones commented out at lines 113-115 *do* display correctly though)